les/distributor.go

```diff
@@ -0,0 +1,333 @@
+// for the Ethereum Light Client.
+package les
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/mclock"
+)
+
+// ErrNoPeers is returned if no peers capable of serving a queued request are
+// available for a certain amount of time (noPeersTimeout)
+var (
+	ErrNoPeers     = errors.New("no suitable peers available")
+	noPeersTimeout = time.Second * 5
+)
+
+// requestDistributor implements a mechanism that distributes requests to
+// suitable peers, obeying flow control rules and prioritizing them in creation
+// order (even when a resend is necessary).
+type requestDistributor struct {
+	queueFirst, queueLast       *distReq
+	lastReqOrder                uint64
+	stopChn                     chan struct{}
+	loopChn                     chan bool
+	loopNextSent, loopRetrySent bool
+	lock                        sync.Mutex
+
+	getAllPeers func() map[distPeer]struct{}
+}
+
+// distPeer is an LES server peer interface for the request distributor.
+// waitBefore returns either the necessary waiting time before sending a request
+// with the given upper estimated cost or the estimated remaining relative buffer
+// value after sending such a request (in which case the request can be sent
+// immediately). At least one of these values is always zero.
+type distPeer interface {
+	waitBefore(uint64) (time.Duration, float64)
+	canSendOrdered() bool
+	orderedSend(f func())
+}
+
+// distReq is the request abstraction used by the distributor. It is based on
+// three callback functions:
+// - getCost returns the upper estimate of the cost of sending the request to a given peer
+// - canSend tells if the server peer is suitable to serve the request
+// - request prepares sending the request to the given peer and returns a function that
+// does the actual sending. Request order should be preserved but the callback itself should not
+// block until it is sent because other peers might still be able to receive requests while
+// one of them is blocking. Instead, the returned function is put in the peer's orderedSend queue.
+type distReq struct {
+	getCost func(distPeer) uint64
+	canSend func(distPeer) bool
+	request func(distPeer) func()
+
+	reqOrder uint64
+	// only for queued requests
+	noPeersTime          mclock.AbsTime
+	queuePrev, queueNext *distReq
+	sentChn              chan distPeer
+}
+
+// newRequestDistributor creates a new request distributor
+func newRequestDistributor(getAllPeers func() map[distPeer]struct{}, stopChn chan struct{}) *requestDistributor {
+	r := &requestDistributor{
+		loopChn:     make(chan bool, 2),
+		stopChn:     stopChn,
+		getAllPeers: getAllPeers,
+	}
+	go r.loop()
+	return r
+}
+
+// newDistReq creates a new request instance that can be queued for distribution
+func newDistReq(getCost func(distPeer) uint64, canSend func(distPeer) bool, request func(distPeer) func()) *distReq {
+	return &distReq{
+		getCost: getCost,
+		canSend: canSend,
+		request: request,
+	}
```

Comment:
Please change callers to use `&distReq{getCost: func(...) uint64 { ... }, ...}` instead of calling newDistReq. Doing this will add "labels" to all the func literals.